presto-hive/src/main/java/com/facebook/presto/hive/CreateEmptyPartitionProcedure.java

```diff
@@ -0,0 +1,123 @@
+            List.class,
+            List.class);
+
+    private final HiveMetadataFactory hiveMetadataFactory;
+    private final ExtendedHiveMetastore metastore;
+    private final LocationService locationService;
+    private final JsonCodec<PartitionUpdate> partitionUpdateJsonCodec;
+
+    @Inject
+    public CreateEmptyPartitionProcedure(HiveMetadataFactory hiveMetadataFactory, ExtendedHiveMetastore metastore, LocationService locationService, JsonCodec<PartitionUpdate> partitionUpdateCodec)
+    {
+        this.hiveMetadataFactory = requireNonNull(hiveMetadataFactory, "hiveMetadataFactory is null");
+        this.metastore = requireNonNull(metastore, "metastore is null");
+        this.locationService = requireNonNull(locationService, "locationService is null");
+        this.partitionUpdateJsonCodec = requireNonNull(partitionUpdateCodec, "partitionUpdateCodec is null");
+    }
+
+    @Override
+    public Procedure get()
+    {
+        return new Procedure(
+                "system",
+                "create_empty_partition",
+                ImmutableList.of(
+                        new Argument("schema", VARCHAR),
+                        new Argument("table", VARCHAR),
+                        new Argument("partitionColumnNames", "array(varchar)"),
+                        new Argument("partitionValues", "array(varchar)")),
+                INSERT_EMPTY_PARTITION.bindTo(this));
+    }
+
+    public void createEmptyPartition(ConnectorSession session, String schema, String table, List<Object> partitionColumnNames, List<Object> partitionValues)
+    {
+        HiveMetadata hiveMetadata = hiveMetadataFactory.create();
+
+        HiveInsertTableHandle hiveInsertTableHandle = hiveMetadata.beginInsert(session, new HiveTableHandle(schema, table));
+
+        List<String> actualPartitionColumnNames = hiveInsertTableHandle.getInputColumns().stream()
+                .filter(HiveColumnHandle::isPartitionKey)
+                .map(HiveColumnHandle::getName)
+                .collect(toImmutableList());
+        checkArgument(Objects.equals(partitionColumnNames, actualPartitionColumnNames), "input partition column names doesn't match actual partition column names");
+
+        List<String> partitionStringValues = partitionValues.stream()
+                .map(String.class::cast)
+                .collect(toImmutableList());
+
+        if (metastore.getPartition(schema, table, partitionStringValues).isPresent()) {
+            throw new PrestoException(ALREADY_EXISTS, "Partition already exists");
+        }
+        String partitionName = FileUtils.makePartName(actualPartitionColumnNames, partitionStringValues);
+
+        WriteInfo writeInfo = locationService.getPartitionWriteInfo(hiveInsertTableHandle.getLocationHandle(), Optional.empty(), partitionName);
+        hiveMetadata.finishInsert(
+                session,
+                hiveInsertTableHandle,
+                ImmutableList.of(
+                    Slices.wrappedBuffer(
```

Comment:
@electrum : Is this hard to read because the purpose of `Slices.wrappedBuffer` is unclear? So the variable name could be something like `serializedPartitionUpdate`